### ISSUE TITLE
image links created by `image-uploader` are blocked by the updated forum filter

### ISSUE THREAD

**Issue by Chiroyce1** [ScratchAddons/ScratchAddons#6913]
### Description

So after the 4-5 days of forum downtime, there seem to be a lot of changes to the post filter, which in turn blocks posting image links created by the `image-uploader` addon.
https://github.com/ScratchAddons/ScratchAddons/blob/f595f3b0f97ca50b8fc18f17674e7105c6e6b696/addons/image-uploader/userscript.js#L139

### Steps to reproduce

1. Try to post with an image uploaded using the `image-uploader` addon

![image](https://github.com/ScratchAddons/ScratchAddons/assets/97374054/066e4cda-9412-4443-b29e-85b4cf8447c2)

*I know this isn't a bug caused by SA, but the functionality of this addon has been hindered.*

**Comment by mybearworld** [comment]
Looks like it's the string `assets.scratch.mit.edu/get_image`.

**Comment by mybearworld** [comment]
> yes, and the forum filter now reflects the behaviour of the filter used in the rest of the site, posting a comment with the _exact_ string got me muted.

Oh wow, that is a *really* weird decision

**Comment by mybearworld** [comment]
> An issue with this though is that it's practically impossible to use images on the forums then, since without get_image, zero images are usable.

Cubeupload still works, I'd imagine.

**Comment by medlans** [comment]
> > An issue with this though is that it's practically impossible to use images on the forums then, since without get_image, zero images are usable.
> 
> Cubeupload still works, I'd imagine.

Yea it does.

> > yes, and the forum filter now reflects the behaviour of the filter used in the rest of the site, posting a comment with the _exact_ string got me muted.
> 
> Oh wow, that is a _really_ weird decision

It might block "scratch.mit.edu/get_image" in general like the mainsite filter. Also, another weird thing I noticed is the DjangoBB filter seems to be stacked onto the main site filter or sth. Try posting something with a word that was already blocked by the DjangoBB filter, and you'll see what I mean and it will replace it with an asterisk instead:  https://scratch.mit.edu/discuss/post/7668163/

**Comment by ajskateboarder** [comment]
I don't really see a fix for this other than creating our own unofficial image host that integrates with Scratch accounts and converts those links to images on the client

**Comment by BroJac5246** [comment]
> > An issue with this though is that it's practically impossible to use images on the forums then, since without get_image, zero images are usable.
> 
> Cubeupload still works, I'd imagine.

It does, though some schools probably block it. I don't know how many education users Scratch has but it's probably a lot.

**Comment by mybearworld** [comment]
Could the addon be replaced to, instead of outputting `[img]https://assets.scratch.mit.edu/get_image/.%2E/d8ca1f3cf01721957c5276fdc3cc683f.png[/img]`, output `[url https://chiroyce1.github.io/image/?dOcaHfJcfGHNIHPLNcLINMfdcJccMOJf.png]image[/url]`?

**Comment by Secret-chest** [comment]
> Could the addon be replaced to, instead of outputting [img]https://assets.scratch.mit.edu/get_image/.%2E/d8ca1f3cf01721957c5276fdc3cc683f.png[/img], output [url https://chiroyce1.github.io/image/?dOcaHfJcfGHNIHPLNcLINMfdcJccMOJf.png]image[/url]?

Please don't make it output links. That should be a setting, otherwise use percent-encoding or the other bypass.

**Comment by Chiroyce1** [comment]
@DNin01 we've already been doing that with the `.%2E` trick, it isn't directly evading the filter but is evading the image host whitelist which is a part of it.

**Comment by mybearworld** [comment]
> we've already been doing that with the `.%2E` trick, it isn't directly evading the filter but is evading the image host whitelist which is a part of it. `https://assets.scratch.mit.edu/{hash}.{extension}` is _actually_ not allowed on image whitelist

Except that this is about the actual filter, to which they (apparently intentionally) added assets. There are a myriad of workarounds, but that would be intentionally breaking Scratch's rules, which is not what Scratch Addons should be.

**Comment by Chiroyce1** [comment]
> which they (apparently intentionally) added assets.

They added it to the main site's filter first since it was a major risk of having a one click link that auto downloads any file, regardless of its format. When it was added to the forums, even though it isn't a clickable link, it still flags it. Instead of saying that it isn't a supported image host it says "unsuitable language" which doesn't seem like intentional behaviour. But all in all, yes, this is on the edge, Paddle has specifically said assets was never meant to be used as an image host.
> it was blocked on the main site because it was being abused in comments and spam attacks.  

[*(source)*](https://scratch.mit.edu/discuss/topic/726667/?page=1#post-7667728)

**Comment by WorldLanguages** [comment]
Are there any other statements by the Scratch Team? (other than paddle2see saying "it was never our intent", which is obvious)

**Comment by Secret-chest** [comment]
> Are there any other statements by the Scratch Team? (other than paddle2see saying "it was never our intent", which is obvious)

I don't think so. It should be safe to fix the links, as they're only a problem on non-forums.

**Comment by Secret-chest** [comment]
Is there any conclusion?

**Comment by WorldLanguages** [comment]
> Is there any conclusion?

I'm not sure. What are forum users doing, now that the addon has been broken for a while? Are they manually adding the bypass? Using cubeupload?

**Comment by ajskateboarder** [comment]
> > Is there any conclusion?
> 
> I'm not sure. What are forum users doing, now that the addon has been broken for a while? Are they manually adding the bypass? Using cubeupload?

I was using the percent-encoding bypasses here and there, but they seem to be patched now

![image](https://github.com/ScratchAddons/ScratchAddons/assets/81255936/2a0c332a-2ce6-4bc7-bb50-24c844ea9988)

**Comment by medlans** [comment]
> > > Is there any conclusion?
> > 
> > 
> > I'm not sure. What are forum users doing, now that the addon has been broken for a while? Are they manually adding the bypass? Using cubeupload?
> 
> I was using the percent-encoding bypasses here and there, but they seem to be patched now
> 
> ![image](https://private-user-images.githubusercontent.com/81255936/291119373-2a0c332a-2ce6-4bc7-bb50-24c844ea9988.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwMDUyMTcsIm5iZiI6MTcwMzAwNDkxNywicGF0aCI6Ii84MTI1NTkzNi8yOTExMTkzNzMtMmEwYzMzMmEtMmNlNi00YmM3LWJiNTAtMjRjODQ0ZWE5OTg4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE5VDE2NTUxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFlYjAyOTYwODhiZjU1MmU3ZjY0MDUzZGY4ZTQ2Y2U1NTkyZmFiYmY0ZmIyM2I5ZDEwZDdmYjlhZDIyMGFiOTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.rnCrbnvvCeQsNSzJARb5Jyiqs58ypfEgKKR7-S6HOV0)

The string “scratch.mit.edu/get_image” itself is blocked.

**Comment by WorldLanguages** [comment]
> I was using the percent-encoding bypasses here and there, but they seem to be patched now

Oh... Does that mean there's no way to use "assets" now?

**Comment by WorldLanguages** [comment]
Looks like we'll need to mark the addon as deprecated.

Any creative ideas to keep the addon useful? Some type integration with cubeupload? In theory, we could add `https://cubeupload.com/*` to the optional host permission of the extension. The end-user would need to register and login manually though.
Also for anyone reading this, it may be a good opportunity to create a general-purpose extension that simplifies uploading images to cubeupload and that could work with multiple types of forums.

**Comment by BroJac5246** [comment]
> Looks like we'll need to mark the addon as deprecated.
> 
> Any creative ideas to keep the addon useful? Some type integration with cubeupload? In theory, we could add `https://cubeupload.com/*` to the optional host permission of the extension. The end-user would need to register and login manually though.
> Also for anyone reading this, it may be a good opportunity to create a general-purpose extension that simplifies uploading images to cubeupload and that could work with multiple types of forums.

I think changing it to Cubeupload is a good idea.

I believe many schools block Cubeupload and I preferred the assets servers but there's not much we can do about those so providing an alternative is probably the best we can do.

**Comment by DNin01** [comment]
If it's possible, CubeUpload integration could maybe work. I'm not familiar with their API(s).

Furthermore, it would also allow you to manage your images easier, whereas with the current assets.scratch.mit.edu method, you can't fully delete images you have uploaded, you just have to not share the URL.

<sub>(...which also brings up the question, should we have noted that images are uploaded to assets.scratch.mit.edu, to cover privacy concerns? I guess we can't do anything about it now that we're deprecating that functionality, at least I think we are)</sub>

But more important privacy-wise then is that we might have to inform users that we're now going to be uploading images to CubeUpload, and do something like temporarily disable the addon until they agree. We'd have to request host permissions anyway, so we won't forget.

IMO, though, I think the most likely course of action I'd take would still be to deprecate the addon entirely.

**Comment by mybearworld** [comment]
> Could the addon be replaced to, instead of outputting `[img]https://assets.scratch.mit.edu/get_image/.%2E/d8ca1f3cf01721957c5276fdc3cc683f.png[/img]`, output `[url https://chiroyce1.github.io/image/?dOcaHfJcfGHNIHPLNcLINMfdcJccMOJf.png]image[/url]`?

This is still an option

**Comment by Secret-chest** [comment]
> This is still an option

No, it's really annoying and depends on @Chiroyce1 's JS.

**Comment by Secret-chest** [comment]
@DNin01 shouldn't have the ultimate decision. Should we do a vote? Paddle2See said it was added there because of the main site issues with downloads, of course he's not going to encourage making more extensions but this is a legitimate use as for some it's the only way to use pictures.

**Comment by mybearworld** [comment]
> > This is still an option
> 
> No, it's really annoying and depends on @Chiroyce1 's JS.

Then it can just link to the raw assets link - I just thought it would be nicer to have that as it wouldn't require a download.

**Comment by Chiroyce1** [comment]
> No, it's really annoying and depends on @Chiroyce1 's JS.

It's better than no image lol - also I think https://scratchaddons.com could have a page with similar functionality to mine.

**Comment by mybearworld** [comment]
> > No, it's really annoying and depends on @Chiroyce1 's JS.
> 
> It's better than no image lol - also I think https://scratchaddons.com could have a page with similar functionality to mine.

We couldn't link it on Scratch then, though.

**Comment by WorldLanguages** [comment]
Uploading images to "assets" and using a custom link that lets you display those images without the browser thinking it needs to download them, doesn't sound too well. It will be blocked by school filters, so at that point just use cubeupload.

> Should we do a vote?

A vote on what? What's the alternative to deprecating the addon? Using links?

**Comment by BroJac5246** [comment]
> > Could the addon be replaced to, instead of outputting `[img]https://assets.scratch.mit.edu/get_image/.%2E/d8ca1f3cf01721957c5276fdc3cc683f.png[/img]`, output `[url https://chiroyce1.github.io/image/?dOcaHfJcfGHNIHPLNcLINMfdcJccMOJf.png]image[/url]`?
> 
> This is still an option

No one actually wants to click a link, though. Also, that is a LOT of images to process.

Additionally, I'm not sure the Scratch Team would be happy. They manually review each new official image host for its ability to moderate images, and we don't have the capacity to do that. Yes, we'd be using their servers, but not wanting to moderate all of that was likely a reason they stopped supporting the assets servers as a host so we shouldn't just implement a workaround that presents the same issue.

> > Should we do a vote?
> 
> A vote on what? What's the alternative to deprecating the addon? Using links?

I think the only valid options are:
1 (🎉). Deprecate the addon and do nothing else
2 (❤️). Deprecate the addon and add a separate Cubeupload (or similar) addon
3 (🚀). Replace the addon with a Cubeupload (or similar) addon.

This would help judge interest, though obviously it wouldn't necessarily be the final decision.

**Comment by OneShot-Niko** [comment]
> 3 (🚀). Replace the addon with a Cubeupload (or similar) addon.

I think it may be better as a different extension rather than addon, however it may be better to replace it like with this option to reduce confusion of "Wait, where did the upload images addon go?"

**Comment by DNin01** [comment]
> Paddle2See said it was added there because of the main site issues with downloads

Wouldn't this problem also apply to the forums?

**Comment by OneShot-Niko** [comment]
> Wouldn't this problem also apply to the forums?

I do remember that in one time a user had left _an assets link_ to a literal text file (could've very well been anything instead but there was room for misuse of assets URLs) which the user had removed themselves later

Then it remained pretty quiet afterwards and the forums didn't have assets be used for anything else as hard as I can remember.

**Comment by DNin01** [comment]
Again, I don't really know how I feel about having an addon that accesses image hosts. Not due to safety/privacy concerns, but just, do we want/need to provide that option in an extension all about Scratch? And it's not like it takes that much longer to upload and attach an image yourself.

<sup>It's hard to explain, I'm sorry.</sup>

There might also be a general-purpose browser extension to do this exact thing on any website.

There are also other image hosts that people might prefer to use, so…

**Comment by DNin01** [comment]
> > Wouldn't this problem also apply to the forums?
> 
> I do remember that in one time a user had left _an assets link_ to a literal text file (could've very well been anything instead but there was room for misuse of assets URLs) which the user had removed themselves later
> 
> Then it remained pretty quiet afterwards and the forums didn't have assets be used for anything else as hard as I can remember.

So it affects the forums. I understand why the team might not want these links on the forums, whether blocking them from being posted on the forums was intentional or not.

**Comment by DNin01** [comment]
Maybe they should fix the downloading vulnerability 😄

**Comment by mybearworld** [comment]
> Maybe they should fix the downloading vulnerability 😄

Haven't they already?

**Comment by Samq64** [comment]
> Replace the addon with a Cubeupload (or similar) addon.

Do any of the whitelisted image hosts have an API though?

**Comment by mybearworld** [comment]
> > Replace the addon with a Cubeupload (or similar) addon.
> 
> Do any of the whitelisted image hosts have an API that doesn't delete images after a while though?

The Scratch Wiki and Wikipedia technically do, I think? I don't think either would appreciate being used an an image host, though

**Comment by CST1229** [comment]
> > Maybe they should fix the downloading vulnerability 😄
> 
> Haven't they already?

I think (at least last time I checked, it now only allows uploading valid images and sounds, though it also allows zip files for some reason).

**Comment by Secret-chest** [comment]
> do we want/need to provide that option in an extension all about Scratch?

Yes. It improves the Scratch experience.

**Comment by BroJac5246** [comment]
> Again, I don't really know how I feel about having an addon that accesses image hosts. Not due to safety/privacy concerns, but just, do we want/need to provide that option in an extension all about Scratch? And it's not like it takes that much longer to upload and attach an image yourself.
> 
> <sup>It's hard to explain, I'm sorry.</sup>

Sure but this is a well-used feature for Scratchers, and many Scratchers primarily use the forums.

> There might also be a general-purpose browser extension to do this exact thing on any website.

There probably is, but it wouldn't be as well-integrated, and most of our users wouldn't seek one out.

> There are also other image hosts that people might prefer to use, so…

...add a setting.

It's like a Mad Libs :P

**Comment by OneShot-Niko** [comment]
> ...add a setting.

Slight issue:
As far as I'm aware, other image hosts besides Cubeupload are paid though.

**Comment by WorldLanguages** [comment]
Well, for next release we will deprecate the addon or somehow remove it from the settings page. In the future, we can consider a sort of cubeupload integration, but no guarantees. I don't think any other alternatives for the future, other than links, were proposed here. Correct me if I'm wrong.

**Comment by Secret-chest** [comment]
Sadly github lacks a 😥 reaction.

**Comment by DNin01** [comment]
We should finish #6987 soon so we can mark this addon as deprecated, ideally for next release, v1.36.0.

**Comment by zaid1442011** [comment]
> > ...add a setting.
> 
> Slight issue: As far as I'm aware, other image hosts besides Cubeupload are paid though.

For some reason, all of my Imageshack images are up and I can upload more, and 30 days passed.

**Comment by OneShot-Niko** [comment]
> For some reason, all of my Imageshack images are up and I can upload more, and 30 days passed.
Pretty sure that can only happen if you are a paid user

**Comment by zaid1442011** [comment]
> > For some reason, all of my Imageshack images are up and I can upload more, and 30 days passed.
> 
> Pretty sure that can only happen if you are a paid user

I didn't pay them. I didn't put a payment method in the first place. Maybe someone else could try this.

**Comment by OneShot-Niko** [comment]
> I didn't pay them. I didn't put a payment method in the first place. Maybe someone else could try this.

I'm probably remembering wrong in that regard.

**Comment by WorldLanguages** [comment]
> We should finish #6987 soon so we can mark this addon as deprecated, ideally for next release, v1.36.0.

For now, I've marked the addon as "beta".

**Comment by Chiroyce1** [comment]
This might be unstable but there is a solution (credits to @jeffalo)
```
[img]https://assets.scratch.mit.edu/w/images/.%2E/.%2E/83ceb675d61ca1fdee0f84c2f5fe5daa.png[/img]
```
- https://scratch.mit.edu/discuss/post/7811380

Can this be added as a temporary fix?

**Comment by mybearworld** [comment]
The Scratch Team has already blocked one such alternative in the past, it's become pretty clear that they don't want images for the forums hosted on assets.scratch.mit.edu for some reason.

**Comment by Chiroyce1** [comment]
Worth a shot in my opinion, of course storing images there isn't a problem for them but I guess it is a concern for moderation, but hey, its a Scratch domain, used by all projects.

**Comment by DNin01** [comment]
@Chiroyce1 Anything can be patched eventually if it becomes a problem. Also, I'd rather not popularize something that could become a safety or security issue in the future.

Again, the issue that caused these URLs to be blocked, as someone else said before, was that apparently assets.scratch.mit.edu was abused to create download links for files, because directly visiting the URL of an asset auto-downloads it. To fix this, the Scratch Team simply blocked those links from comments on Scratch, via the filter bot. Recently, when the forums were updated, this rule was carried over to forum posts. It was likely an unintentional side effect, but assets.scratch.mit.edu was never intended to be used as an image host for Scratch forum posts anyway, plus, there was already another rule that prevented these links from being postable on the forums without inserting `/get_image/.%2E`.

But that's just my opinion, and to be fair, we have already been bypassing filters through this addon. However, the addon [has already been functionally deprecated](https://github.com/ScratchAddons/ScratchAddons/commits/f4d5eab1142cfa1accd2c98ac46e9e5b0c5b681e/addons/image-uploader) as of v1.36.0, so we don't really have any choice other than to not support it anymore, unless we integrate with third-party image hosts like CubeUpload.

**Comment by Secret-chest** [comment]
`https://%61ssets.scratch.mit.edu/scratchr2/static/.%2E/.%2E/` works.

> assets.scratch.mit.edu was never intended to be used as an image host for Scratch forum posts anyway

then why is it on the whitelist

We can randomise which letters are percent-encoded.

**Comment by mybearworld** [comment]
> then why is it on the whitelist

It's not. `/get_image/.%2E` is a hack to get around the whitelist.

**Comment by Chiroyce1** [comment]
@WorldLanguages I think its best we reach a conclusion soon 
*(pinged you since you reacted to one of the workarounds with a rocket so I thought maybe you still consider this? idk)*

We could *maybe* add one of these variations of % encodings and non existent paths to get the image working or close this issue for good, its been a good 4 years of this revolutionary addon that allowed so many more forumers to use images without hosting on a different site :)
<img width="402" alt="image" src="https://github.com/ScratchAddons/ScratchAddons/assets/97374054/e8bc5d50-4b30-4c86-a274-10b42c014176">

**Comment by OneShot-Niko** [comment]
> then why is it on the whitelist

the whitelist probably has `scratch.mit.edu/*` on it, definitely for a reason, but not one we know probably

**Comment by CST1229** [comment]
> the whitelist probably has `scratch.mit.edu/*` on it, definitely for a reason, but not one we know probably

I think I once figured out the whitelist has `*.scratch.mit.edu` on it, but only if the pathname also contains `get_image/` or `scratchr2/static/`. So basically 2.0 site asset URLs and uploads/cdn2 thumbnails match by default, but you can use any other image on any scratch.mit.edu domain with the .%2E/ trick.

**Comment by ninjamar** [comment]
What about using the asset link without get_image. Eg `https://assets.scratch.mit.edu/HASH`

**Comment by CST1229** [comment]
> What about using the asset link without get_image. Eg `https://assets.scratch.mit.edu/HASH`

That won't pass the image link whitelist. Assets images work because all `scratch.mit.edu` domain images with `get_image/` in the path are whitelisted. I assume this was made for project/user/studio thumbnails but using some path tricks (`get_image/.%2E`) you can get it to work for any domain.

**Comment by WorldLanguages** [comment]
#7661 removes the addon and closes this issue.

Please create another issue to discuss CubeUpload integration. Note that it's still quite out-of-scope, and it's weird to provide a feature that needs a website account to work, but we should still track it somewhere.
